### DISCUSSION TITLE
Does cgroup system.slice/k3s.service reflect the resource usage of the k3s service itself?

### DISCUSSION THREAD

**Discussion by codeReaper2001** [k3s-io/k3s#11470] [Q&A]
Hi everyone，I am currently considering making a monitoring dashboard for k3s-server/k3s-agent resource usage, but now I have some doubts.

I noticed that k3s is hosted to run in systemd and kubelet (including cadvisor) is integrated into the k3s binary, kubelet startup parameters are also generated during the k3s initialization phase and then passed to the kubelet program, in the code I found this part of the code：
https://github.com/k3s-io/k3s/blob/b55aaeb08d4d54efce39eca0d475e4080b1e177c/pkg/daemons/agent/agent_linux.go#L150-L163
This should be why I was able to get the cadvisor metrics with the id k3s.service
<img width="1208" alt="image" src="https://github.com/user-attachments/assets/a29980fd-dd3d-4088-a06e-944a3e5638ef" />

In addition, I found that the resource usage test of the k3s official website also uses the corresponding indicators here
![image](https://github.com/user-attachments/assets/2635cf45-d643-4139-9bc9-c297c3ac0880)

However, when I used the systemd-cgls command to view the cgroup details of k3s, I found many child processes, including containerd. I want to know:
(1) When were these child processes added to the k3s cgroup? I did not find the relevant code yet
(2) Can the container_xxx metrics above accurately describe the resource usage of k3s-server/k3s-agent?

```bash
root@mounto:/sys/fs/cgroup/system.slice/k3s.service# systemd-cgls
Working directory /sys/fs/cgroup/system.slice/k3s.service:
├─   6290 /var/lib/rancher/k3s/data/0d5a433a75956a27ce0e2513f50eaaee76c6055b93ceda9f4047d2eb73054f3e/bin/containerd-shim-runc-v2 -namespace k8s.io -id d>
├─   6427 /var/lib/rancher/k3s/data/0d5a433a75956a27ce0e2513f50eaaee76c6055b93ceda9f4047d2eb73054f3e/bin/containerd-shim-runc-v2 -namespace k8s.io -id b>
├─   6683 /var/lib/rancher/k3s/data/0d5a433a75956a27ce0e2513f50eaaee76c6055b93ceda9f4047d2eb73054f3e/bin/containerd-shim-runc-v2 -namespace k8s.io -id 5>
├─   6836 /var/lib/rancher/k3s/data/0d5a433a75956a27ce0e2513f50eaaee76c6055b93ceda9f4047d2eb73054f3e/bin/containerd-shim-runc-v2 -namespace k8s.io -id b>
├─   6997 /var/lib/rancher/k3s/data/0d5a433a75956a27ce0e2513f50eaaee76c6055b93ceda9f4047d2eb73054f3e/bin/containerd-shim-runc-v2 -namespace k8s.io -id a>
├─   7087 /var/lib/rancher/k3s/data/0d5a433a75956a27ce0e2513f50eaaee76c6055b93ceda9f4047d2eb73054f3e/bin/containerd-shim-runc-v2 -namespace k8s.io -id a>
├─   7385 /var/lib/rancher/k3s/data/0d5a433a75956a27ce0e2513f50eaaee76c6055b93ceda9f4047d2eb73054f3e/bin/containerd-shim-runc-v2 -namespace k8s.io -id 6>
...
├─3233693 containerd
...
```

**Comment by dereknola** [comment]
(1) This code is pretty dense, but you can see [here](https://github.com/k3s-io/k3s/blob/master/pkg/containerd/main.go#L14) how we internally launch containerd. 
(2) Yes, they can. K3s runs an embedded version of containerd. The cgroup slice is giving you everything that constitutes the binary download of K3s (containerd, the controlplane components, and kubelet (for agents) etcd).
